components/TextFields/src/MDCTextInputController.m

```diff
@@ -0,0 +1,1201 @@
+    return;
+  }
+
+  if (object == _textInput.leadingUnderlineLabel &&
+      ![_textInput.leadingUnderlineLabel.font isEqual:[[self class] underlineLabelsFont]]) {
+    _customLeadingFont = _textInput.leadingUnderlineLabel.font;
+  } else if (object == _textInput.placeholderLabel &&
+             ![_textInput.placeholderLabel.font isEqual:[[self class] placeholderFont]]) {
+    _customPlaceholderFont = _textInput.placeholderLabel.font;
+  } else if (object == _textInput.trailingUnderlineLabel &&
+             ![_textInput.trailingUnderlineLabel.font isEqual:[[self class] underlineLabelsFont]]) {
+    _customTrailingFont = _textInput.trailingUnderlineLabel.font;
+  } else {
+    return;
+  }
+  [self updateLayout];
+}
+
+#pragma mark - Public API
+
+- (void)setErrorText:(NSString *)errorText
+    errorAccessibilityValue:(NSString *)errorAccessibilityValue {
+  // Turn on error:
+  //
+  // Here the 'magic' logic happens for error text.
+  // When the user sets error text, we save the current state of their underline, leading text,
+  // trailing text, and placeholder text for both content and color.
+  if (errorText && !self.isDisplayingErrorText) {
+    // If we are not in error, but will be, we need to save the existing state.
+    self.previousLeadingText = self.textInput.leadingUnderlineLabel.text
+                                   ? self.textInput.leadingUnderlineLabel.text.copy
+                                   : @"";
+
+    self.textInput.leadingUnderlineLabel.text = errorText;
+  }
+
+  // Change error:
+  if (errorText && self.isDisplayingErrorText) {
+    self.textInput.leadingUnderlineLabel.text = errorText;
+  }
+
+  // Turn off error:
+  //
+  // If error text is unset (nil) we reset to previous values.
+  if (!errorText) {
+    // If there is a saved state, use it.
+    self.textInput.leadingUnderlineLabel.text =
+        self.previousLeadingText ?: self.textInput.leadingUnderlineLabel.text;
+
+    // Clear out saved state.
+    self.previousLeadingText = nil;
+  }
+
+  self.errorText = errorText;
+  self.errorAccessibilityValue = errorAccessibilityValue;
+
+  [self updateLayout];
+
+  // Accessibility
+  if (errorText) {
+    NSString *announcementString = errorAccessibilityValue;
+    if (!announcementString.length) {
+      announcementString =
+          errorText.length > 0 ? [NSString stringWithFormat:@"Error: %@", errorText] : @"Error.";
+    }
+
+    // Simply sending a layout change notification does not seem to
+    UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, announcementString);
+
+    NSString *valueString = @"";
+
+    if (self.textInput.text > 0) {
+      valueString = [self.textInput.text copy];
+    }
+    if (self.textInput.placeholder.length > 0) {
+      valueString =
+          [NSString stringWithFormat:@"%@. %@.", valueString, self.textInput.placeholder.copy];
+    }
+    valueString = [NSString stringWithFormat:@"%@.", valueString];
```

Comment:
This will result in a double '..' at the end of the string if there is textInput.placeholder.
Consider moving this before the if clause just above, or:
valueString = [NSString stringWithFormat:"%@.", self.textInput.text]